syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.workers";
option java_outer_classname = "RouteGuideProto";
option objc_class_prefix = "RTG";

package proto;

// Interface exported by the server.
service Employees
{
  // A simple RPC.
  rpc GetEmployee(EmployeeId) returns (Employee) {}
  rpc InsertEmployee(Employee) returns (EmployeeId) {}
  rpc GetEmployeePosition(EmployeeId) returns (EmployeePosition) {}
  rpc GetEmployeeManager(EmployeeId) returns (Manager) {}
  rpc SetEmployeePosition(SetEmployeePositionRequest) returns (SetEmployeePositionResponce) {}
  rpc SetEmployeeManager(SetEmployeeManagerRequest) returns (SetEmployeeManagerResponce) {}
  rpc DeleteEmployee(EmployeeId) returns (DeleteEmployeeResponce) {}

  // A server-to-client streaming RPC.
  rpc ListEmployees(EmptyMessage) returns (stream Employee) {}
}

message Employee
{
  uint64 id = 1;
  optional uint64 manager_id = 2;
  string name = 3;
  string position = 4;
}

message EmployeeId { uint64 id = 1; }

message Manager { optional Employee employee = 1; }

message EmptyMessage {}

message EmployeePosition { string position = 1; }

message SetEmployeePositionRequest
{
  uint64 id = 1;
  string position = 2;
}
message SetEmployeePositionResponce {}

message SetEmployeeManagerRequest
{
  uint64 id = 1;
  uint64 manager_id = 2;
}

message SetEmployeeManagerResponce {}

message DeleteEmployeeResponce {}